File: src/main/java/org/jukeboxmc/world/generator/noise/SimplexOctaveGenerator.java
package org.jukeboxmc.world.generator.noise;

import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.world.generator.noise.bukkit.NoiseGenerator;

import java.util.Arrays;
import java.util.Random;

public class SimplexOctaveGenerator extends PerlinOctaveGenerator {

    public SimplexOctaveGenerator(@NotNull Random rand, int octaves, int sizeX, int sizeZ ) {
        this( rand, octaves, sizeX, 1, sizeZ );
    }

    public SimplexOctaveGenerator(@NotNull Random rand, int octaves, int sizeX, int sizeY, int sizeZ ) {
        super( createOctaves( rand, octaves ), rand, sizeX, sizeY, sizeZ );
    }

    public SimplexOctaveGenerator(@NotNull Random rand, int octaves ) {
        this( rand, octaves, 0, 0, 0 );
    }

    protected static NoiseGenerator @NotNull [] createOctaves(@NotNull Random rand, int octaves ) {
        NoiseGenerator[] result = new NoiseGenerator[octaves];

        for ( int i = 0; i < octaves; i++ ) {
            result[i] = new SimplexNoise( rand );
        }

        return result;
    }

    @Override
    public double[] getFractalBrownianMotion( double x, double y, double z, double lacunarity, double persistence ) {
        Arrays.fill( noise, 0 );

        double freq = 1;
        double amp = 1;

        // fBm
        for ( NoiseGenerator octave : octaves ) {
            noise = ( (PerlinNoise) octave ).getNoise( noise, x, y, z, sizeX, sizeY, sizeZ, xScale * freq, yScale * freq, zScale * freq, 0.55D / amp );
            freq *= lacunarity;
            amp *= persistence;
        }

        return noise;
    }
}

File: src/main/java/org/jukeboxmc/world/generator/noise/SimplexOctaveGenerator.kt
package org.jukeboxmc.world.generator.noise

import org.jukeboxmc.world.generator.noise.bukkit.NoiseGenerator
import java.util.Arrays
import java.util.Random

class SimplexOctaveGenerator(
    rand: Random,
    octaves: Int,
    sizeX: Int = 0,
    sizeY: Int = 0,
    sizeZ: Int = 0,
) : PerlinOctaveGenerator(
    createOctaves(rand, octaves),
    rand,
    sizeX,
    sizeY,
    sizeZ,
) {
    constructor(rand: Random, octaves: Int, sizeX: Int, sizeZ: Int) : this(rand, octaves, sizeX, 1, sizeZ)

    override fun getFractalBrownianMotion(
        x: Double,
        y: Double,
        z: Double,
        lacunarity: Double,
        persistence: Double,
    ): DoubleArray? {
        Arrays.fill(noise, 0.0)
        var freq = 1.0
        var amp = 1.0

        // fBm
        for (octave in octaves) {
            noise = (octave as PerlinNoise).getNoise(
                noise,
                x,
                y,
                z,
                sizeX,
                sizeY,
                sizeZ,
                xScale * freq,
                yScale * freq,
                zScale * freq,
                0.55 / amp,
            )
            freq *= lacunarity
            amp *= persistence
        }
        return noise
    }

    companion object {
        protected fun createOctaves(rand: Random, octaves: Int): Array<NoiseGenerator> {
            return Array(octaves) {
                SimplexNoise(rand)
            }
        }
    }
}
